File: R/nat.fastcore/tests/testthat/test-warp.R
# Landmark transforms: thin-plate spline and moving least squares.
#
# There is no external binary to check against here (unlike CMTK/Elastix), so these pin the
# maths on properties that hold for any correct implementation: both transforms interpolate
# their landmarks, and both reproduce a global affine exactly when the landmarks are related
# by one. The Python test suite additionally checks us against `morphops` and `molesq`, the
# reference implementations `navis` delegates to.

set.seed(42)

# A non-degenerate landmark pair with a genuinely non-affine component.
n_lm <- 40L
src <- matrix(runif(n_lm * 3, 0, 100), ncol = 3)
trg <- src %*% diag(c(1.3, 0.8, 1.1)) +
  matrix(c(5, -3, 12), n_lm, 3, byrow = TRUE) +
  matrix(rnorm(n_lm * 3, 0, 4), ncol = 3)

pts <- matrix(runif(200 * 3, 0, 100), ncol = 3)

# Landmarks related by a pure affine, for the exact-recovery checks.
aff_linear <- rbind(c(2, 0.1, 0), c(0, -1, 0.25), c(0.3, 0, 0.75))
aff_offset <- c(3, 4, 5)
src_aff <- matrix(runif(20 * 3, 0, 10), ncol = 3)
apply_affine <- function(m) m %*% aff_linear + matrix(aff_offset, nrow(m), 3, byrow = TRUE)
trg_aff <- apply_affine(src_aff)


# ---------------------------------------------------------------------------
# Thin-plate spline
# ---------------------------------------------------------------------------

test_that("tps interpolates its landmarks", {
  tps <- tps_transform(src, trg)
  expect_equal(tps_xform(tps, src), trg, tolerance = 1e-6)
})

test_that("tps recovers a global affine, including far outside the landmark hull", {
  tps <- tps_transform(src_aff, trg_aff)
  far <- matrix(runif(50 * 3, -500, 500), ncol = 3)
  expect_equal(tps_xform(tps, far), apply_affine(far), tolerance = 1e-6)
})

test_that("tps_affine is the homogeneous form of that affine", {
  m <- tps_affine(tps_transform(src_aff, trg_aff))
  expect_equal(dim(m), c(4L, 4L))
  expect_equal(m[4, ], c(0, 0, 0, 1))
  # The linear part is applied as a row-vector product, so the matrix form is its transpose.
  expect_equal(m[1:3, 1:3], t(aff_linear), tolerance = 1e-8)
  expect_equal(m[1:3, 4], aff_offset, tolerance = 1e-8)
})

test_that("fitting the other way round inverts the mapping", {
  back <- tps_transform(trg, src)
  expect_equal(tps_xform(back, trg), src, tolerance = 1e-6)
})

test_that("tps_coefs returns the fit", {
  co <- tps_coefs(tps_transform(src, trg))
  expect_equal(dim(co$source), c(n_lm, 3L))
  expect_equal(dim(co$W), c(n_lm, 3L))
  expect_equal(dim(co$A), c(4L, 3L))
  expect_equal(co$source, src)
})

test_that("a singular system is an R error, not a panic or nonsense", {
  # Duplicate source landmarks pulled to different targets cannot be interpolated.
  bad_src <- rbind(c(0, 0, 0), c(0, 0, 0), c(1, 0, 0), c(0, 1, 0), c(0, 0, 1))
  bad_trg <- rbind(c(0, 0, 0), c(5, 0, 0), c(1, 0, 0), c(0, 1, 0), c(0, 0, 1))
  expect_error(tps_transform(bad_src, bad_trg), "singular")
})

test_that("tps needs at least four landmarks", {
  p <- rbind(c(0, 0, 0), c(1, 0, 0), c(0, 1, 0))
  expect_error(tps_transform(p, p), "at least 4")
})

test_that("tps print method reports the landmark count", {
  expect_output(print(tps_transform(src, trg)), "40 landmarks")
})


# ---------------------------------------------------------------------------
# Moving least squares
# ---------------------------------------------------------------------------

test_that("mls interpolates its landmarks", {
  mls <- mls_transform(src, trg)
  expect_equal(mls_xform(mls, src), trg, tolerance = 1e-6)
})

test_that("mls recovers a global affine, including far outside the landmark hull", {
  mls <- mls_transform(src_aff, trg_aff)
  far <- matrix(runif(50 * 3, -500, 500), ncol = 3)
  expect_equal(mls_xform(mls, far), apply_affine(far), tolerance = 1e-6)
})

test_that("mls_affine is the homogeneous form of that affine", {
  m <- mls_affine(mls_transform(src_aff, trg_aff))
  expect_equal(dim(m), c(4L, 4L))
  expect_equal(m[4, ], c(0, 0, 0, 1))
  expect_equal(m[1:3, 1:3], t(aff_linear), tolerance = 1e-8)
  expect_equal(m[1:3, 4], aff_offset, tolerance = 1e-8)
})

test_that("direction = 'inverse' maps targets back to sources", {
  back <- mls_transform(src, trg, direction = "inverse")
  expect_equal(mls_xform(back, trg), src, tolerance = 1e-6)
  expect_equal(back$direction, "inverse")
})

test_that("inverting is not the same transform as the forward one", {
  fwd <- mls_transform(src, trg)
  inv <- mls_transform(src, trg, direction = "inverse")
  expect_false(isTRUE(all.equal(mls_xform(fwd, pts), mls_xform(inv, pts))))
})

test_that("mls rejects an unknown direction", {
  expect_error(mls_transform(src, trg, direction = "sideways"))
})

test_that("mls print method reports landmarks and direction", {
  expect_output(print(mls_transform(src, trg)), "40 landmarks, direction 'forward'")
})

test_that("mls handles landmark counts the reference implementation cannot", {
  # molesq builds (3, M, N) intermediates; this shape is out of reach for it but is only
  # the output matrix here.
  big_src <- matrix(runif(2000 * 3, 0, 1000), ncol = 3)
  big_trg <- big_src + matrix(rnorm(2000 * 3, 0, 5), ncol = 3)
  out <- mls_xform(mls_transform(big_src, big_trg), pts)
  expect_equal(dim(out), dim(pts))
  expect_true(all(is.finite(out)))
})


# ---------------------------------------------------------------------------
# Shared input handling
# ---------------------------------------------------------------------------

builders <- list(
  tps = list(build = tps_transform, xform = tps_xform),
  mls = list(build = mls_transform, xform = mls_xform)
)

for (nm in names(builders)) {
  b <- builders[[nm]]

  test_that(paste(nm, "accepts a length-3 vector and data frames"), {
    tr <- b$build(src, trg)
    expect_equal(b$xform(tr, c(50, 50, 50)), b$xform(tr, matrix(c(50, 50, 50), ncol = 3)))

    df_src <- as.data.frame(src)
    names(df_src) <- c("x", "y", "z")
    df_trg <- as.data.frame(trg)
    names(df_trg) <- c("X", "Y", "Z") # capitalised, as `nat` writes them
    expect_equal(b$xform(b$build(df_src, df_trg), pts), b$xform(tr, pts))
  })

  test_that(paste(nm, "rejects mismatched landmark counts"), {
    expect_error(b$build(src, trg[-1, ]), "must match")
  })

  test_that(paste(nm, "rejects non-3-column input"), {
    expect_error(b$build(src[, 1:2], trg[, 1:2]), "\\(N, 3\\)")
  })

  test_that(paste(nm, "rejects non-finite landmarks"), {
    bad <- src
    bad[1, 1] <- NA
    expect_error(b$build(bad, trg), "NA, NaN or Inf")
  })

  test_that(paste(nm, "rejects a foreign object"), {
    expect_error(b$xform(list(), pts), "must be")
  })

  test_that(paste(nm, "thread count does not change the result"), {
    tr <- b$build(src, trg)
    expect_identical(b$xform(tr, pts, n_cores = 1), b$xform(tr, pts))
    expect_identical(b$xform(tr, pts, n_cores = 4), b$xform(tr, pts))
  })
}

File: py/tests/test_warp.py
"""Landmark-based transforms: thin-plate spline and moving least squares.

Where `navis` is importable these check us against it directly - `navis.transforms`
delegates to `morphops` and `molesq`, which are the reference implementations. Where it is
not, the properties checked here (landmark interpolation, exact recovery of a global
affine) pin the maths on their own.
"""

import pickle

import numpy as np
import pytest

import navis_fastcore as fastcore

# The reference implementations agree with us to ~1e-14 *relative to the spatial extent of
# the landmarks*. These tolerances are on data spanning ~100 units, so they leave several
# orders of magnitude of headroom. A change that pushes agreement to 1e-6 means something is
# subtly wrong - do not loosen them.
ATOL = 1e-9


@pytest.fixture(scope="module")
def landmarks():
    """A non-degenerate landmark pair with a genuinely non-affine component."""
    rng = np.random.default_rng(42)
    src = rng.uniform(0, 100, size=(40, 3))
    trg = src * [1.3, 0.8, 1.1] + [5.0, -3.0, 12.0]
    trg += rng.normal(0, 4.0, size=src.shape)  # the warp the affine cannot explain
    return src, trg


@pytest.fixture(scope="module")
def points():
    rng = np.random.default_rng(7)
    return rng.uniform(0, 100, size=(500, 3))


CLASSES = [fastcore.TpsTransform, fastcore.MlsTransform]
IDS = ["tps", "mls"]


# ---------------------------------------------------------------------------
# Against navis / the reference implementations
# ---------------------------------------------------------------------------


def _rel_err(got, ref, src):
    """Max deviation relative to the spatial extent of the landmarks."""
    return np.abs(got - ref).max() / np.linalg.norm(src.max(axis=0) - src.min(axis=0))


def test_tps_matches_morphops(landmarks, points):
    """The whole point: same numbers as the implementation we are replacing.

    Compared against `morphops` rather than `navis.transforms.TPStransform`, because that
    class *is* these four lines - and importing `navis` drags in its full optional
    dependency stack, which makes the check flaky for reasons unrelated to the maths.
    """
    mops = pytest.importorskip("morphops", reason="morphops not installed")
    # morphops still calls `np.row_stack`, removed in numpy 2. navis hotfixes this the same
    # way; without it the reference cannot run at all on a current numpy.
    if not hasattr(np, "row_stack"):
        np.row_stack = np.vstack

    src, trg = landmarks
    W, A = mops.tps_coefs(src, trg)
    ref = mops.P_matrix(points) @ A + mops.K_matrix(points, src) @ W

    got = fastcore.TpsTransform(src, trg).xform(points)
    assert _rel_err(got, ref, src) < 1e-10


def test_mls_matches_molesq(landmarks, points):
    """Same, for moving least squares: `molesq` is what `navis` delegates to."""
    molesq = pytest.importorskip("molesq", reason="molesq not installed")
    src, trg = landmarks

    ref = molesq.Transformer(src, trg).transform(points)
    got = fastcore.MlsTransform(src, trg).xform(points)
    assert _rel_err(got, ref, src) < 1e-10


def test_mls_reverse_matches_molesq(landmarks, points):
    """`direction="inverse"` must mean what `molesq`'s `reverse=True` means."""
    molesq = pytest.importorskip("molesq", reason="molesq not installed")
    src, trg = landmarks

    ref = molesq.Transformer(src, trg).transform(points, reverse=True)
    got = fastcore.MlsTransform(src, trg, direction="inverse").xform(points)
    assert _rel_err(got, ref, src) < 1e-10


# ---------------------------------------------------------------------------
# Properties that hold regardless of implementation
# ---------------------------------------------------------------------------


@pytest.mark.parametrize("cls", CLASSES, ids=IDS)
def test_landmarks_map_to_targets(landmarks, cls):
    """Both transforms are interpolating: a source landmark lands on its partner."""
    src, trg = landmarks
    got = cls(src, trg).xform(src)
    assert np.allclose(got, trg, atol=1e-6)


@pytest.mark.parametrize("cls", CLASSES, ids=IDS)
def test_recovers_global_affine(cls):
    """Affine landmarks in, that exact affine out - everywhere, including far outside."""
    rng = np.random.default_rng(3)
    src = rng.uniform(0, 10, size=(20, 3))

    def affine(v):
        return v @ np.array([[2.0, 0.1, 0.0], [0.0, -1.0, 0.25], [0.3, 0.0, 0.75]]) + [
            3.0,
            4.0,
            5.0,
        ]

    trg = affine(src)
    tr = cls(src, trg)

    pts = rng.uniform(-500, 500, size=(50, 3))  # well outside the landmark hull
    assert np.allclose(tr.xform(pts), affine(pts), atol=1e-6)


@pytest.mark.parametrize("cls", CLASSES, ids=IDS)
def test_matrix_affine(cls):
    """`matrix_affine` is the homogeneous form of the recovered global affine."""
    rng = np.random.default_rng(11)
    src = rng.uniform(0, 10, size=(20, 3))
    linear = np.array([[2.0, 0.1, 0.0], [0.0, -1.0, 0.25], [0.3, 0.0, 0.75]])
    offset = np.array([3.0, 4.0, 5.0])
    trg = src @ linear + offset

    m = cls(src, trg).matrix_affine
    assert m.shape == (4, 4)
    assert np.allclose(m[3], [0, 0, 0, 1])
    # `linear` is applied as a row-vector product, so the matrix form is its transpose.
    assert np.allclose(m[:3, :3], linear.T, atol=1e-8)
    assert np.allclose(m[:3, 3], offset, atol=1e-8)


# ---------------------------------------------------------------------------
# Input handling
# ---------------------------------------------------------------------------


@pytest.mark.parametrize("cls", CLASSES, ids=IDS)
def test_single_point_roundtrips_as_1d(landmarks, cls):
    src, trg = landmarks
    tr = cls(src, trg)
    out = tr.xform(np.array([50.0, 50.0, 50.0]))
    assert out.shape == (3,)
    assert np.allclose(out, tr.xform(np.array([[50.0, 50.0, 50.0]]))[0])


@pytest.mark.parametrize("cls", CLASSES, ids=IDS)
def test_dataframe_input(landmarks, points, cls):
    """DataFrames with x/y/z columns work for both landmarks and points."""
    pd = pytest.importorskip("pandas")
    src, trg = landmarks
    src_df = pd.DataFrame(src, columns=["x", "y", "z"])
    trg_df = pd.DataFrame(trg, columns=["x", "y", "z"])
    pts_df = pd.DataFrame(points, columns=["x", "y", "z"])

    assert np.allclose(cls(src_df, trg_df).xform(pts_df), cls(src, trg).xform(points))


@pytest.mark.parametrize("cls", CLASSES, ids=IDS)
def test_dataframe_missing_columns(landmarks, cls):
    pd = pytest.importorskip("pandas")
    src, trg = landmarks
    bad = pd.DataFrame(src, columns=["x", "y", "not_z"])
    with pytest.raises(ValueError, match="x/y/z columns"):
        cls(bad, trg)


@pytest.mark.parametrize("cls", CLASSES, ids=IDS)
def test_rejects_mismatched_landmarks(landmarks, cls):
    src, trg = landmarks
    with pytest.raises(ValueError, match="must match"):
        cls(src, trg[:-1])


@pytest.mark.parametrize("cls", CLASSES, ids=IDS)
def test_rejects_non_3d(cls):
    bad = np.zeros((10, 2))
    with pytest.raises(ValueError, match=r"\(N, 3\)"):
        cls(bad, bad)


def test_tps_needs_four_landmarks():
    pts = np.array([[0.0, 0, 0], [1, 0, 0], [0, 1, 0]])
    with pytest.raises(ValueError, match="at least 4 landmarks"):
        fastcore.TpsTransform(pts, pts)


def test_mls_rejects_bad_direction(landmarks):
    src, trg = landmarks
    with pytest.raises(ValueError, match="forward"):
        fastcore.MlsTransform(src, trg, direction="sideways")


# ---------------------------------------------------------------------------
# Threading
# ---------------------------------------------------------------------------


@pytest.mark.parametrize("cls", CLASSES, ids=IDS)
@pytest.mark.parametrize("n_cores", [1, 2, 4])
def test_thread_count_does_not_change_result(landmarks, points, cls, n_cores):
    """Chunking is over independent points, so results must be bit-identical."""
    src, trg = landmarks
    tr = cls(src, trg)
    assert np.array_equal(tr.xform(points, n_cores=n_cores), tr.xform(points))


@pytest.mark.parametrize("cls", CLASSES, ids=IDS)
def test_empty_input(landmarks, cls):
    src, trg = landmarks
    out = cls(src, trg).xform(np.zeros((0, 3)))
    assert out.shape == (0, 3)


# ---------------------------------------------------------------------------
# Object protocol
# ---------------------------------------------------------------------------


@pytest.mark.parametrize("cls", CLASSES, ids=IDS)
def test_len_and_repr(landmarks, cls):
    src, trg = landmarks
    tr = cls(src, trg)
    assert len(tr) == len(src)
    assert "landmarks=40" in repr(tr)


@pytest.mark.parametrize("cls", CLASSES, ids=IDS)
def test_equality(landmarks, cls):
    src, trg = landmarks
    assert cls(src, trg) == cls(src, trg)
    assert cls(src, trg) != cls(src, trg * 2)
    assert cls(src, trg) != "not a transform"


@pytest.mark.parametrize("cls", CLASSES, ids=IDS)
def test_copy(landmarks, points, cls):
    src, trg = landmarks
    tr = cls(src, trg)
    assert np.array_equal(tr.copy().xform(points), tr.xform(points))


@pytest.mark.parametrize("cls", CLASSES, ids=IDS)
def test_negation_inverts_direction(landmarks, cls):
    """Negation maps targets back onto sources. Both are refits, not true inverses."""
    src, trg = landmarks
    back = (-cls(src, trg)).xform(trg)
    assert np.allclose(back, src, atol=1e-6)


@pytest.mark.parametrize("cls", CLASSES, ids=IDS)
def test_pickle_roundtrip(landmarks, points, cls):
    src, trg = landmarks
    tr = cls(src, trg)
    restored = pickle.loads(pickle.dumps(tr))
    assert np.allclose(restored.xform(points), tr.xform(points))
    assert len(restored) == len(tr)


def test_tps_pickle_does_not_refit(landmarks):
    """The coefficients travel with the pickle, so unpickling is exact, not a re-solve."""
    src, trg = landmarks
    tr = fastcore.TpsTransform(src, trg)
    restored = pickle.loads(pickle.dumps(tr))
    assert np.array_equal(restored.W, tr.W)
    assert np.array_equal(restored.A, tr.A)


def test_mls_direction_and_negation(landmarks, points):
    """`direction="inverse"` and negating a forward transform are the same thing."""
    src, trg = landmarks
    inverse = fastcore.MlsTransform(src, trg, direction="inverse")
    negated = -fastcore.MlsTransform(src, trg)
    assert inverse.direction == "inverse" == negated.direction
    assert np.array_equal(inverse.xform(points), negated.xform(points))
    assert np.array_equal(inverse.source, trg)
    assert np.array_equal(inverse.target, src)


# ---------------------------------------------------------------------------
# TPS coefficients
# ---------------------------------------------------------------------------


def test_tps_from_coefs(landmarks, points):
    """Rebuilding from W/A reproduces the transform without refitting."""
    src, trg = landmarks
    tr = fastcore.TpsTransform(src, trg)
    rebuilt = fastcore.TpsTransform.from_coefs(tr.source, tr.W, tr.A)
    assert np.array_equal(rebuilt.xform(points), tr.xform(points))
    assert rebuilt.target is None


def test_tps_from_coefs_matches_numpy_fit(landmarks, points):
    """A caller may fit with LAPACK and hand us the coefficients; same answer either way."""
    src, trg = landmarks
    m = len(src)

    # Build and solve the saddle-point system the way morphops does.
    k = np.linalg.norm(src[:, None, :] - src[None, :, :], axis=-1)
    p = np.column_stack([np.ones(m), src])
    L = np.zeros((m + 4, m + 4))
    L[:m, :m] = k
    L[:m, m:] = p
    L[m:, :m] = p.T
    rhs = np.vstack([trg, np.zeros((4, 3))])
    q = np.linalg.solve(L, rhs)

    rebuilt = fastcore.TpsTransform.from_coefs(src, q[:m], q[m:])
    assert np.allclose(rebuilt.xform(points), fastcore.TpsTransform(src, trg).xform(points),
                       atol=ATOL)


def test_tps_from_coefs_validates_shapes(landmarks):
    src, trg = landmarks
    tr = fastcore.TpsTransform(src, trg)
    with pytest.raises(ValueError, match=r"`W` must be"):
        fastcore.TpsTransform.from_coefs(src, tr.W[:-1], tr.A)
    with pytest.raises(ValueError, match=r"`A` must be"):
        fastcore.TpsTransform.from_coefs(src, tr.W, tr.A[:-1])


def test_tps_from_coefs_cannot_negate_without_target(landmarks):
    src, trg = landmarks
    tr = fastcore.TpsTransform(src, trg)
    rebuilt = fastcore.TpsTransform.from_coefs(src, tr.W, tr.A)
    with pytest.raises(ValueError, match="landmarks_target"):
        -rebuilt


def test_tps_coefficient_shapes(landmarks):
    src, trg = landmarks
    tr = fastcore.TpsTransform(src, trg)
    assert tr.W.shape == (len(src), 3)
    assert tr.A.shape == (4, 3)
    assert np.array_equal(tr.source, src)
    assert np.array_equal(tr.target, trg)


def test_tps_singular_landmarks_raise():
    """Duplicate landmarks pulling apart is unsolvable - say so, don't return nonsense."""
    src = np.array([[0.0, 0, 0], [0, 0, 0], [1, 0, 0], [0, 1, 0], [0, 0, 1]])
    trg = np.array([[0.0, 0, 0], [5, 0, 0], [1, 0, 0], [0, 1, 0], [0, 0, 1]])
    with pytest.raises(ValueError, match="singular"):
        fastcore.TpsTransform(src, trg)


# ---------------------------------------------------------------------------
# Scale
# ---------------------------------------------------------------------------


def test_large_landmark_count_is_tractable():
    """The reference MLS cannot run this shape at all; peak memory here is the output."""
    rng = np.random.default_rng(1)
    src = rng.uniform(0, 1000, size=(2000, 3))
    trg = src + rng.normal(0, 5, size=src.shape)
    pts = rng.uniform(0, 1000, size=(5000, 3))

    out = fastcore.MlsTransform(src, trg).xform(pts)
    assert out.shape == pts.shape
    assert np.isfinite(out).all()
